### ISSUE TITLE
Error: 'MySQL' is not recognised as an internal or external command

### ISSUE THREAD

**Issue by RobertLutken** [resin-io/etcher#1025]
![mysqlerror_etcher](https://cloud.githubusercontent.com/assets/5775654/21930725/5be4911e-d98f-11e6-9e72-f753ec2fa509.PNG)
- 1.0.0-beta 17
- Windows 10 Pro 64-bit

**Comment by lurch** [comment]
We've previously seen something very similar to this on Gitter https://gitter.im/resin-io/etcher?at=582a14e945c9e3eb4339d229
I didn't realise we didn't have a Github issue for it yet.

@jviotti Did we ever get any ideas what might be causing this?

@RobertLutken Would you be willing to help us debug this? Do you currently have MySQL installed on your PC, or have you had it installed previously?

**Comment by RobertLutken** [comment]
@lurch I do indeed have MySQL installed, well MySQL workbench. 
A little bit of digging seems to refer to "A key is a string specified in the Add method that uniquely identifies a specific member of a collection." [MSDN](https://msdn.microsoft.com/en-us/library/aa232626(v=vs.60).aspx)
I'll have a dig around and try setup some different environment variables. 
Happy to help!

**Comment by jviotti** [comment]
Hey there @RobertLutken !

Thanks a lot for reporting! We've very recently fixed the `Microsoft VBScript runtime error: This key is already associated with an
element of this collection` error (see https://github.com/resin-io-modules/drivelist/pull/130).

Regarding the MySQL error, I think its simply output coming from your environment every time you type a command, and should be safely ignored by our drive detection parser (see https://github.com/resin-io-modules/drivelist/pull/99).

So just to confirm the issue is gone, can you try the following `bat` scripts on your computer, and report back the results?

- After the fix: https://github.com/resin-io-modules/drivelist/blob/master/scripts/win32.bat
- Before the fix: https://github.com/resin-io-modules/drivelist/blob/cdfc551e5fbe874a09cc947589480548bd938712/scripts/win32.bat

Also, can you open `cmd.exe` and try typing some commands on it? I want to check if those MySQL warnings you get get printed every time you write a command there, independently on what they are. You can try `wmic diskdrive list brief` for instance.

**Comment by daturach** [comment]
Hello

Same error message, in French this time  :-)

Etcher-1.0.0-beta.18-win32-x64
Windows 7, 64 bits Edition Familiale Premium

Anything I can do to help?

**Comment by jviotti** [comment]
Hi @daturach ,

Can you try running the `.bat` scripts I linked to in my previous comment?

Also, what happens if you *any* command on `cmd.exe`? Do you get the MySQL warning as well?

**Comment by jviotti** [comment]
For this user, it seems to happen everytime a `cmd.exe` window is opened. Maybe its the same thing?

https://i.stack.imgur.com/e8b8q.png

**Comment by jviotti** [comment]
Hi @daturach ,

Can you also paste a screenshot of the error dialog?

**Comment by daturach** [comment]
Hello @jviotti 

Thanks for your quick answer.
This is the error message I get when starting etcher:
![etcher_mysql_error](https://cloud.githubusercontent.com/assets/9865568/22365037/a728b268-e475-11e6-8be8-1a0767870722.png)


As a side note, this dialog box is not even supporting accented chars...:-(


Running win32.bat before the fix:
![before_bat](https://cloud.githubusercontent.com/assets/9865568/22365196/6fe0ee32-e476-11e6-99ce-0a6e6edb68d9.png)

This MySQL error message is displayed only once. Other commands run into the same cmd.exe does not show this message anymore.

Running win32.bat after the fix displays the list of devices as above but without the error message if run in the same session.

Command `wmic diskdrive list brief` shows list of devices but never displays the error message.

Hope that helps!
Thanks

**Comment by jviotti** [comment]
I think I have a theory about this. While our YAML parser will ignore invalid keys (https://github.com/resin-io-modules/drivelist/pull/99), the JavaScript code that executes the `.bat` script as a child process will blow up with an error if the `stderr` from the child process is not empty:

```js
// See https://github.com/resin-io-modules/drivelist/blob/master/lib/scripts.js
if (stderr.trim().length) {
    return callback(new Error(stderr));
}
```

This seems to be what's happening given that the error contains *both* MySQL warning instead of just one (if the YAML parser would have been complaining about this, it would have complained about the first warning only, and we would have a YAML related stack trace).

I'm not very sure about what to do here. We can of course ignore `stderr` completely, but we might be hiding real errors accidentally, which is definitely not good. 

@lurch What do you think?

**Comment by jviotti** [comment]
In the mid-time, in order to validate this theory, I'll produce a custom Windows build that logs a warning instead of throwing an error if the child process includes `stderr` output.

**Comment by lurch** [comment]
Hmmm, given that we've _only_ ever seen this error for MySQL (for a still-unknown reason) but for multiple times, I'd be tempted to filter out only any stderr messages starting with `'MySQL' ` which matches both the English and French errors above.

Also, should there be a separate issue for the French error-dialog not correctly showing accented characters?

**Comment by jviotti** [comment]
> Hmmm, given that we've only ever seen this error for MySQL (for a still-unknown reason)

It seems to be a MySQL `%PATH%` configuration issue. Try googling the warning, the error seems to be well known. The reason why we see it here is that when you execute `cmd.exe`, it tries to run some MySQL command by default, and thus we get it on every new `cmd.exe` instance, including the one that triggers the drive detection script.

> I'd be tempted to filter out only any stderr messages starting with 'MySQL' which matches both the English and French errors above.

Yeah, maybe, although the fact that a similar error might be reported in the future makes me feel a bit uncomfortable.

I have an alternative solution: What about allowing the drivelist client to pass a function to handler `stderr` from the scripts? We can throw by default, but Etcher could silently print them in DevTools (and we'll see it if a user has any problems and we ask him to send us the DevTools output), and send them to TrackJS.

A similar approach would be to allow some option in drivelist to print `stderr` output to `stderr` instead of throwing (basically proxying the output). Since in Etcher we send any `stderr` output to TrackJS, we accomplish the same thing automatically.

**Comment by lurch** [comment]
> It seems to be a MySQL %PATH% configuration issue. Try googling the warning, the error seems to be well known.

I googled, and as expected the first few results are from people actually trying to run `mysql` (which we aren't). But then I found https://github.com/cmderdev/cmder/issues/443 which has lots of great info, especially https://github.com/cmderdev/cmder/issues/443#issuecomment-150202124 which suggests that using double-quotes is the way to fix this problem, i.e. maybe we simply need to change https://github.com/resin-io-modules/drivelist/blob/master/scripts/win32.bat#L6 to
```
set "PATH=%PATH%;%SYSTEMROOT%\System32"
```
:-D

**Comment by daturach** [comment]
@lurch, jviotti

Running win32.bat with the PATH between "" does not produce any MySQL error.

**Comment by jviotti** [comment]
Amazing discovery! This might indeed be the reason! I'm uploading a custom build with these changes and we'll see how it goes :)

**Comment by jviotti** [comment]
@daturach Can you try the following custom build as well just to confirm? https://drive.google.com/open?id=0B7tkbonGU-RyNTRxdlBhRVgwQnc.

**Comment by daturach** [comment]
Thanks
What's the process? unzip and run Etcher.exe?
I get the following:
![etcher_test1](https://cloud.githubusercontent.com/assets/9865568/22386552/e0161a40-e4d7-11e6-88b4-f2d11aa5e30e.png)

**Comment by jviotti** [comment]
Let me try here. I might be messed something up when editing the files straight in the asar archive.

**Comment by jviotti** [comment]
Try this one: https://drive.google.com/open?id=0B7tkbonGU-RyUnJxbm5WeVpSM1U. Indeed messed up the previous one, sorry about that :)

**Comment by daturach** [comment]
Ok thanks; it's looking fine. MySQL error message disappeared.
Next week, I will start flashing some images. Great work!

**Comment by jviotti** [comment]
Woohoo! Thanks a lot for the help on this issue. I'll send a patch to apply this fix!

**Comment by jviotti** [comment]
All the credit to @lurch and his amazing observations of course \o/

**Comment by alexandrosm** [comment]
should we just be double-quoting everything-in-general and get it over
with? it seems so many issues end with "use double quotes"

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

**Comment by jviotti** [comment]
I'll definitely check if we can use double quotes elsewhere. At least in the Windows script, everything else is quoted already.

**Comment by jviotti** [comment]
I've sent two PRs addressing issues highlighted by `shellcheck`, a shell script linter, for both the GNU/Linux and OS X scripts:

- https://github.com/resin-io-modules/drivelist/pull/134
- https://github.com/resin-io-modules/drivelist/pull/135

This tool, among other improvements, highlighted all the places where we should quote for safety purposes.
